### PR TITLE
coredumpctl: implement --since/--until (-S/-U) for info/list verbs

### PR DESCRIPTION
**Pull request by GiedriusS**
journalctl has these two useful options for filtering the output entries so I think the same kind of funcionality would be nice in coredumpctl as well as it is just a "front-end" to journalctl. Played around a bit and it seems like I got the logic right. Please let me know if you have any comments or should I fix the code style because it is my first "real" contribution to systemd.

### CODE REVIEW

File: src/coredump/coredumpctl.c [hunk 1]
```diff
@@ -50,6 +50,9 @@
 #include "user-util.h"
 #include "util.h"
 
+static usec_t arg_since, arg_until;
+static bool arg_since_set, arg_until_set;
```
Comment (poettering) [inline src/coredump/coredumpctl.c]: the booleans should be initialized explicitly to false.

That said, I think it would make sense to remove the booleans and simply initialize arg_since and arg_until to the special value USEC_INFINITY, and then check explicitly for that.

File: src/coredump/coredumpctl.c [hunk 2]
```diff
@@ -128,11 +131,12 @@ static void help(void) {
                "     --no-pager      Do not pipe output into a pager\n"
                "     --no-legend     Do not print the column headers.\n"
                "  -1                 Show information about most recent entry only\n"
+               "  -S --since=DATE    Only print core dumps since the date\n"
+               "  -U --until=DATE    Only print core dumps until the date\n"
```
Comment (poettering) [inline src/coredump/coredumpctl.c]: the help text so far spelt that as "coredump", not "core dump" (i.e. no space), and I figure we should stick to one spelling.

File: src/coredump/coredumpctl.c [hunk 3]
```diff
@@ -610,18 +638,60 @@ static int dump_list(sd_journal *j) {
...
+
+                for (;;) {
+                        if (!arg_reverse)
+                                r = sd_journal_next(j);
+                        else
+                                r = sd_journal_previous(j);
+
+                        if (r < 0) {
+                                log_error_errno(r, "Failed to iterate through journal: %m");
+                                return r;
```
Comment (poettering) [inline src/coredump/coredumpctl.c]: you can merge these two lines into one:

```c
if (r < 0)
        return log_error_errno(r, "Failed to iterate through journal: %m");
```

File: src/coredump/coredumpctl.c [hunk 4]
```diff
@@ -610,18 +638,60 @@ static int dump_list(sd_journal *j) {
...
+                                break;
+
+                        if (arg_until_set && !arg_reverse) {
+                                usec_t usec;
+
+                                r = sd_journal_get_realtime_usec(j, &usec);
+                                if (r < 0) {
+                                        log_error_errno(r, "Failed to determine timestamp: %m");
                                         return r;
+                                }
```
Comment (poettering) [inline src/coredump/coredumpctl.c]: similar as above, you can make the if block a single line

File: src/coredump/coredumpctl.c [hunk 5]
```diff
@@ -610,18 +638,60 @@ static int dump_list(sd_journal *j) {
...
-                                if (r < 0)
+
+                        if (arg_since_set && arg_reverse) {
+                                usec_t usec;
+
+                                r = sd_journal_get_realtime_usec(j, &usec);
+                                if (r < 0) {
+                                        log_error_errno(r, "Failed to determine timestamp: %m");
                                         return r;
+                                }
```
Comment (poettering) [inline src/coredump/coredumpctl.c]: here as well.

**Comment by GiedriusS** [comment]
Changed the commit according to your comments:
* Shortened the if blocks where possible
* Changed 'core dump' to 'coredump' in the usage string
* Removed the two booleans, added a explicit initialization of arg_since/arg_until to USEC_INFINITY and now the code checks for that instead of booleans

Also added two commits to amend the zsh/bash autocompletion scripts and add information about the new options to man/coredumpctl XML files.

File: src/coredump/coredumpctl.c [hunk 6]
```diff
@@ -197,6 +201,22 @@ static int parse_argv(int argc, char *argv[]) {
...
+                        if (r < 0) {
+                                log_error("Failed to parse timestamp: %s", optarg);
+                                return -EINVAL;
+                        }
+                        break;
+
+                case 'U':
+                        r = parse_timestamp(optarg, &arg_until);
+                        if (r < 0) {
+                                log_error("Failed to parse timestamp: %s", optarg);
```
Comment (benjarobin) [inline src/coredump/coredumpctl.c]: I think we should indicate in the error message which option (-S or -U) we failed to parse

Comment (poettering) [reply src/coredump/coredumpctl.c]: well, the offending optarg is shown in the message, hence I think this is clear enough.

File: src/coredump/coredumpctl.c [hunk 7]
```diff
@@ -197,6 +201,22 @@ static int parse_argv(int argc, char *argv[]) {
 
                         break;
 
+                case 'S':
+                        r = parse_timestamp(optarg, &arg_since);
+                        if (r < 0) {
+                                log_error("Failed to parse timestamp: %s", optarg);
+                                return -EINVAL;
+                        }
```
Comment (poettering) [inline src/coredump/coredumpctl.c]: you could shorten this to this btw:

```
r = parse_timestamp(optarg, &arg_since);
if (r < 0)
        return log_error_errno(r, "Failed … );
```

That's not only shorter, but also propagates the error number to the logs and the parent...

Comment (GiedriusS) [reply src/coredump/coredumpctl.c]: Pushed updated commit with this suggested change. I guess journalctl could use these improvements as well because at the moment -EINVAL is returned in these identical cases.
